### PR TITLE
Npm2100 driver

### PR DESCRIPTION
**Pull request by nordic-auko**
Add drivers for nPM2100 PMIC.
This is a multi-function device with regulators, gpio, watchdog, adc, and other supporting functions.

### CODE REVIEW

**Review by WalkingTalkingPotato — APPROVED** [review]

**Comment by martinjaeger** [comment]
@nordic-auko this needs a rebase.

Your commits state "Initial driver for...". What does the "initial" mean? Is it still a draft or incomplete implementation? If not, I'd suggest to remove it to avoid any confusion.

**Comment by nordic-auko** [comment]
> @nordic-auko this needs a rebase.
> 
> Your commits state "Initial driver for...". What does the "initial" mean? Is it still a draft or incomplete implementation? If not, I'd suggest to remove it to avoid any confusion.

Rebased it now.
The commit messages were based on the npm1300 pmic driver. I've changed the wording now to be more concise: it's not a draft or incomplete driver.

File: dts/bindings/watchdog/nordic,npm2100-wdt.yaml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+# Copyright (c) 2024 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+description: nPM2100 Watchdog
+
+compatible: "nordic,npm2100-wdt"
```
Comment (martinjaeger) [inline dts/bindings/watchdog/nordic,npm2100-wdt.yaml:6]: AFAIK `base.yaml` should be included for all drivers.

```suggestion
compatible: "nordic,npm2100-wdt"

include: base.yaml
```

File: drivers/watchdog/wdt_npm2100.c [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+	const struct wdt_npm2100_config *config = dev->config;
+
+	if (channel_id != 0) {
+		return -EINVAL;
+	}
+
+	return i2c_reg_write_byte_dt(&config->i2c, TIMER_TASKS_KICK, 1U);
+}
+
+static const struct wdt_driver_api wdt_npm2100_api = {
```
Comment (martinjaeger) [inline drivers/watchdog/wdt_npm2100.c:116]: This should follow the new approach introduced in #82300.
```suggestion
static DEVICE_API(wdt, wdt_npm2100_api) = {
```

**Review by martinjaeger — CHANGES_REQUESTED** [review]
Only reviewed the watchdog part, where I got a minor change request.

File: drivers/mfd/Kconfig.npm2100 [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+	select I2C
+	help
+	  Enable the Nordic nPM2100 PMIC multi-function device driver
+
+config MFD_NPM2100_LEVEL_IRQ
+	bool "nPM2100 PMIC IRQ level"
+	depends on MFD_NPM2100
+	help
+	  Use level triggering instead of edge triggering when pin interrupt is enabled.
+	  This can reduce power consumption on some platforms, at the expense of irq latency.
```
Comment (gmarull) [inline drivers/mfd/Kconfig.npm2100:17]: shouldn't this be in DT?

File: drivers/mfd/mfd_npm2100.c [hunk 4]
```diff
@@ -0,0 +1,425 @@
...
+	}
+
+	ret = gpio_pin_configure_dt(&config->host_int_gpios, GPIO_INPUT);
+	if (ret < 0) {
+		return ret;
+	}
+
+	gpio_init_callback(&mfd_data->gpio_cb, gpio_callback, BIT(config->host_int_gpios.pin));
+
+	ret = gpio_add_callback(config->host_int_gpios.port, &mfd_data->gpio_cb);
```
Comment (gmarull) [inline drivers/mfd/mfd_npm2100.c:282]: nit: use `gpio_add_callback_dt`

File: drivers/gpio/gpio_npm2100.c [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+static int gpio_npm2100_pin_interrupt_configure(const struct device *dev, gpio_pin_t pin,
+						enum gpio_int_mode mode, enum gpio_int_trig trig)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(pin);
+	ARG_UNUSED(mode);
+	ARG_UNUSED(trig);
+
+	return -ENOTSUP;
+}
```
Comment (gmarull) [inline drivers/gpio/gpio_npm2100.c:157]: can be deleted

File: drivers/gpio/gpio_npm2100.c [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+	.pin_interrupt_configure = gpio_npm2100_pin_interrupt_configure,
+};
+
+static int gpio_npm2100_init(const struct device *dev)
+{
+	const struct gpio_npm2100_config *config = dev->config;
+
+	if (!i2c_is_ready_dt(&config->i2c)) {
+		return -ENODEV;
+	}
```
Comment (gmarull) [inline drivers/gpio/gpio_npm2100.c:175]: question: shouldn't register access go through MFD, so calls are _atomic_?

File: drivers/sensor/nordic/npm2100_vbat/npm2100_vbat.c [hunk 7]
```diff
@@ -0,0 +1,563 @@
...
+			return -EINVAL;
+		}
+
+		adc_cfg->enabled = val->val1 == 0 ? false : true;
+
+		return 0;
+	}
+
+	/* Delay of vbat ADC measurement */
+	if ((chan == SENSOR_CHAN_GAUGE_VOLTAGE) &&
```
Comment (gmarull) [inline drivers/sensor/nordic/npm2100_vbat/npm2100_vbat.c:363]: wondering if this should be a fuel gauge driver?

**Review by gmarull — COMMENTED** [review]
good stuff, a few comments from my side.

File: dts/bindings/watchdog/nordic,npm2100-wdt.yaml (hunk 1 above)
Comment (gmarull) [reply dts/bindings/watchdog/nordic,npm2100-wdt.yaml:6]: I guess only if they use any base.yaml property

Comment (martinjaeger) [reply dts/bindings/watchdog/nordic,npm2100-wdt.yaml:6]: Yeah, it uses at least the `reg` property from `base.yaml`.

Comment (nordic-auko) [reply dts/bindings/watchdog/nordic,npm2100-wdt.yaml:6]: It doesn't use any properties other than what is inherited from the mfd parent device. I don't know if that involves base.yaml.

File: drivers/mfd/Kconfig.npm2100 (hunk 3 above)
Comment (nordic-auko) [reply drivers/mfd/Kconfig.npm2100:17]: Yeah, I was thinking about that, but wasn't sure exactly how to frame it. I'll add it as an enum property or bool then. As far as I understand, the GPIO DT flags doesn't include the interrupt flags

File: drivers/gpio/gpio_npm2100.c (hunk 6 above)
Comment (nordic-auko) [reply drivers/gpio/gpio_npm2100.c:175]: I don't fully understand. Do you mean having a i2c wrapper in the mfd driver, and protect that with a mutex?

File: drivers/sensor/nordic/npm2100_vbat/npm2100_vbat.c (hunk 7 above)
Comment (nordic-auko) [reply drivers/sensor/nordic/npm2100_vbat/npm2100_vbat.c:363]: The npm2100 hardware only measures voltages, which I think fits better as a sensor device driver. To provide more full fledged fuel gauge functionality, additional software is required. We do have this kind of software in NCS/sdk-nrfxlib, and we are considering supporting the zephyr fuel gauge API, but I think that is a larger task outside the scope of this driver

File: dts/bindings/watchdog/nordic,npm2100-wdt.yaml (hunk 1 above)
Comment (martinjaeger) [reply dts/bindings/watchdog/nordic,npm2100-wdt.yaml:6]: You are right. Should be ok not to import base.yaml.

**Review by martinjaeger — DISMISSED** [review]
Watchdog part OK.

**Review by gmarull — APPROVED** [review]

**Review by martinjaeger — APPROVED** [review]

**Pull request merged**